bin/ex_libris_holdings_xml_parser.rb

```diff
@@ -0,0 +1,195 @@
+        @record_count += 1
+        ht_record = HTRecord.new(marc_record)
+        yield ht_record
+      end
+    end
+  end
+end
+
+# A HTRecord has a marc record and knows how to turn it into a tsv string.
+class HTRecord
+  def initialize(marc_record)
+    @marc_record = marc_record # ... a Marc::Record!
+  end
+
+  # Should use the same order as HTRecord.header_tsv
+  def to_tsv
+    [item_type, oclc, local_id, status, condition, enum_chron, issn, govdoc]
+      .join("\t").delete("\n")
+  end
+
+  def to_mon_tsv
+    [oclc, local_id, status, condition, enum_chron, govdoc]
+      .join("\t").delete("\n")
+  end
+
+  def to_ser_tsv
+    [oclc, local_id, issn, govdoc]
+      .join("\t").delete("\n")
+  end
+
+  def itm(x)
+    unless @marc_record["ITM"]
+      raise ArgumentError, "Record rejected for lacking an ITM field: #{@marc_record}"
+    end
+
+    @marc_record["ITM"][x]
+  end
+
+  def leader
+    @leader ||= @marc_record.leader
+  end
+
+  def oclc
+    if @marc_record["035"]
+      @oclc = @marc_record["035"]["a"].strip
+    else
+      raise ArgumentError, "Record rejected for missing 035a (ocn): #{@record}"
+    end
+  end
+
+  def local_id
+    @local_id ||= (item_type == "ser" ? @marc_record["001"].value : itm("d")).strip
+  end
+
+  def condition
+    @condition ||= itm("c")
+  end
+
+  def item_type
+    @item_type ||= map_item_type(leader[7])
+  end
+
+  # ExLibris say: ITM|a: volume, ITM|b: issue, ITM|i: year, ITM|j: month
+  def enum_chron
+    @enum_chron ||= [
+      itm("a"), # volume
+      itm("b"), # issue
+      itm("i"), # year
+      itm("j") ## month
+    ].reject{ |x| x.nil? || x.empty? }.map(&:strip).join(",")
+  end
+
+  def status
+    @status ||= map_status(itm("k"))
+  end
+
+  def issn
+    if item_type == "ser"
+      if @marc_record.fields("022").any?
+        @marc_record["022"]["a"]
+      else
+        []
+      end
+    end
+  end
+
+  # TODO? double triple quadruple check that 17 and 28 are correct and using the right index (0/1)
+  def govdoc
+    @govdoc ||= is_us_govdoc? ? '1' : '0'
+  end
+
+  def is_us_govdoc?
+    unless @marc_record["008"]
+      raise ArgumentError, "Record rejected for missing 008 field: #{@marc_record}"
+    end
+
+    str_val = @marc_record["008"].value.downcase
+
+    pubplace_008 = str_val[15,3]
+    govpub_008 = str_val[28]
+
+    is_us?(pubplace_008) && govpub_008 == 'f'
+  end
+
+  # via post-zephir processing "clean_pub_place"
+  def is_us?(pub_place)
+    return true if pub_place[2] == 'u'
+    return true if pub_place[0,2] == 'pr'
+    return true if pub_place[0,2] == 'us'
+    return false
+  end
+
+  private
+
+  def map_item_type(item_type)
+    {
+      "s" => "ser",
+      "m" => "mon",
+    }[item_type] || "mix"
+  end
+
+  def map_status(status)
+    unless item_type == "ser"
+      {
+        "MISSING" => "LM",
+        "LOST_LOAN"  => "LM",
+      }[status] || "CH"
+    end
+  end
+end
+
+# Parse any incoming files, output to stdout.
+if $0 == __FILE__
```

Comment:
Depending on how/if we plan to integrate this further, I would consider moving ARGV out of the initializer and putting it here -- have the initializer take institution and files parameters, would make it easier to test. That, however, can be a TODO.

Reply:
Yes, will do.